src/runtime/synchronization_common.h

```diff
@@ -571,35 +571,34 @@ struct validate_action {
     uintptr_t invalid_unpark_info = 0;
 };
 
-WEAK bool parking_control_validate(void *control, validate_action &action) {
-    return true;
-};
-WEAK void parking_control_before_sleep(void *control){};
-WEAK uintptr_t parking_control_unpark(void *control, int unparked, bool more_waiters) {
-    return 0;
-};
-WEAK void parking_control_requeue_callback(void *control, const validate_action &action, bool one_to_wake, bool some_requeued){};
-
 struct parking_control {
-    bool (*validate)(void *control, validate_action &action);
-    void (*before_sleep)(void *control);
-    uintptr_t (*unpark)(void *control, int unparked, bool more_waiters);
-    void (*requeue_callback)(void *control, const validate_action &action, bool one_to_wake, bool some_requeued);
-
-    ALWAYS_INLINE parking_control()
-        : validate(parking_control_validate), before_sleep(parking_control_before_sleep),
-          unpark(parking_control_unpark), requeue_callback(parking_control_requeue_callback) {
+    uintptr_t park(uintptr_t addr);
+    uintptr_t unpark_one(uintptr_t addr);
+    int unpark_requeue(uintptr_t addr_from, uintptr_t addr_to, uintptr_t unpark_info);
+
+protected:
+    virtual bool validate(validate_action &action) {
+        return true;
+    }
+    virtual void before_sleep() {
+        // nothing
+    }
+    virtual uintptr_t unpark(int unparked, bool more_waiters) {
+        return 0;
+    }
+    virtual void requeue_callback(const validate_action &action, bool one_to_wake, bool some_requeued) {
+        // nothing
```

Comment:
Do we ever want to create a base `parking_control`? If not, it might make sense to disable construction by adding a pure virtual destructor here.
```
    virtual ~parking_control() = 0;

protected:
    // ...
};

inline parking_control::~parking_control() = default;
```

Reply:
nah, but I'm not worries about it being misused in this way

Reply:
Cool :)